### PR TITLE
8275318: loaded_classes_do may see ArrayKlass before InstanceKlass is loaded

### PR DESCRIPTION
**Pull request by iklam**
When a shared class K is restored, all of its archived array classes K[], K[][], etc, are also restored as well. Such restored classes are placed in the ClassLoaderData::_klasses list.

For safety, ClassLoaderData::loaded_classes_do() should skip all these classes until K is marked as "loaded".

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8275318](https://bugs.openjdk.java.net/browse/JDK-8275318): loaded_classes_do may see ArrayKlass before InstanceKlass is loaded

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to e17ddbf764243a62c79f01796410f5449d6aaff2
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7069/head:pull/7069` \
`$ git checkout pull/7069`

Update a local copy of the PR: \
`$ git checkout pull/7069` \
`$ git pull https://git.openjdk.java.net/jdk pull/7069/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7069`

View PR using the GUI difftool: \
`$ git pr show -t 7069`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7069.diff">https://git.openjdk.java.net/jdk/pull/7069.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/classfile/classLoaderData.cpp [hunk 1]
```diff
@@ -358,15 +359,28 @@ void ClassLoaderData::loaded_classes_do(KlassClosure* klass_closure) {
...
   for (Klass* k = Atomic::load_acquire(&_klasses); k != NULL; k = k->next_link()) {
-    // Do not filter ArrayKlass oops here...
-    if (k->is_array_klass() || (k->is_instance_klass() && InstanceKlass::cast(k)->is_loaded())) {
+    // Filter out InstanceKlasses (or their ObjArrayKlasses) that have not entered the
+    // loaded state.
+    if (k->is_instance_klass()) {
+      if (!InstanceKlass::cast(k)->is_loaded()) {
+        continue;
+      }
+    } else if (k->is_objArray_klass()) {
```
Comment (coleenp) [inline src/hotspot/share/classfile/classLoaderData.cpp:368]: I was wondering if it would help performance to also check is_shared() ?

**Review by coleenp — APPROVED** [review]
Looks good. Just a small comment.

**Review by calvinccheung — APPROVED** [review]
LGTM

**Comment by iklam** [comment]
Thanks to @coleenp and @calvinccheung for the review.
/integrate

**Pull request closed without merging**